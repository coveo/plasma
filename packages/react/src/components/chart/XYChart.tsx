import * as d3 from 'd3';
import {FunctionComponent, createContext} from 'react';
import * as _ from 'underscore';

import {ChartBaseProps} from './ChartContainer';
import {ChartUtils} from './ChartUtils';

export interface XYPoint {
    x: number;
    y: number;
}

export interface XYSerie {
    label: string;
    data: XYPoint[];
}

export type Padding = {top?: number; right?: number; bottom?: number; left?: number};

export interface XYChartProps extends ChartBaseProps {
    series: XYSerie[];
    padding?: number | Padding;
    colorPattern?: string[];
    color?: (serie: number, colorPattern: string[], point?: XYPoint) => string;
    xFormat?: (value: number) => string | number;
    yFormat?: (value: number) => string | number;
    xTicksCount?: number;
    yTicksCount?: number;
}

export interface XYChartContextProps {
    width: number;
    height: number;
    xDomain: [number, number];
    yDomain: [number, number];
    series: XYSerie[];
    colorPattern: string[];
    xScale: d3.scale.Ordinal<number, number>;
    yScale: d3.scale.Linear<number, number>;
    xFormat: (value: number) => string | number;
    yFormat: (value: number) => string | number;
    xTicksCount: number;
    yTicksCount: number;
    color?: (serie: number, colorPattern: string[], point?: XYPoint) => string;
}

export const XYChartContext = createContext<XYChartContextProps>({
    width: 0,
    height: 0,
    xDomain: [0, 0],
    yDomain: [0, 0],
    series: [],
    colorPattern: [],
    xScale: d3.scale.ordinal<number, number>(),
    yScale: d3.scale.linear(),
    xFormat: _.identity,
    yFormat: _.identity,
    xTicksCount: 10,
    yTicksCount: 10,
});

const getDateChartColorPattern = (numOfColors: number) => {
    const colorPattern: string[] = [
        'var(--deprecated-azure)',
        'var(--deprecated-turquoise)',
        'var(--deprecated-anakiwa)',
        'var(--deprecated-nepal)',
    ];
    return numOfColors > 2 ? ['var(--deprecated-blue-8)', ...colorPattern] : colorPattern;
};
const defaultPadding = 10;

/**
 * @deprecated Use Mantine instead
 */
export const XYChart: FunctionComponent<XYChartProps> = ({
    width,
    height,
    series,
    colorPattern = [],
    color,
    padding = defaultPadding,
    xFormat = _.identity,
    yFormat = _.identity,
    xTicksCount,
    yTicksCount,
    children,
}) => {
    const xValues = ChartUtils.getXValues(series);
    const xDomain: [number, number] = [Math.min(...xValues), Math.max(...xValues)];

    const yValues = ChartUtils.getYValues(series);
    const yDomain: [number, number] = [Math.min(0, ...yValues), Math.max(...yValues)];

    colorPattern = colorPattern.length ? colorPattern : getDateChartColorPattern(series.length);
    color = color || ((serie: number, pattern: string[]) => pattern[serie]);

    const chartPadding: Padding = _.isNumber(padding)
        ? {top: padding, right: padding, bottom: padding, left: padding}
        : _.defaults(padding, {
              top: defaultPadding,
              right: defaultPadding,
              bottom: defaultPadding,
              left: defaultPadding,
          });

    width = width - chartPadding.left - chartPadding.right;
    height = height - chartPadding.top - chartPadding.bottom;

    const xScale = d3.scale.ordinal<number, number>().domain(xValues).rangePoints([0, width]);

    const yScale = d3.scale.linear<number, number>().domain(yDomain).range([height, 0]);

    return (
        <XYChartContext.Provider
            value={{
                width,
                height,
                xDomain,
                yDomain,
                xScale,
                yScale,
                xFormat,
                yFormat,
                color,
                colorPattern,
                series,
                xTicksCount: xTicksCount || xValues.length / 2,
                yTicksCount: yTicksCount || yValues.length / 2,
            }}
        >
            <svg width={width} height={height}>
                <g transform={`translate(${chartPadding.left},${chartPadding.top})`}>{children}</g>
            </svg>
        </XYChartContext.Provider>
    );
};
XYChart.displayName = 'XYChart';
