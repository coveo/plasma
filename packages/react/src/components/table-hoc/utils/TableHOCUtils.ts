import * as _ from 'underscore';

import {isValidElement} from 'react';
import {PlasmaState} from '../../../PlasmaState';
import {DatePickerSelectors} from '../../datePicker/DatePickerSelectors';
import {IFilterState} from '../../filterBox/FilterBoxReducers';
import {FlatSelectSelectors} from '../../flatSelect/FlatSelectSelectors';
import {IListBoxState} from '../../listBox/ListBoxReducers';
import {PaginationUtils} from '../../pagination/PaginationUtils';
import {ITableRowOwnProps} from '../TableRowConnected';
import {ITableWithSortState} from '../reducers/TableWithSortReducers';

export interface ITableHOCPredicateValue {
    id: string;
    value: string;
}

export interface ITableHOCCompositeState {
    sortKey?: string;
    sortAscending?: boolean;
    perPage?: number;
    pageNb?: number;
    filter?: string;
    predicates?: ITableHOCPredicateValue[];
    dateLimits?: [Date, Date?];
}

const PREDICATE_SEPARATOR = '--';

const getCompositeState = (id: string, state: PlasmaState): ITableHOCCompositeState => {
    const tableSort: ITableWithSortState = _.find(
        state.tableHOCHeader,
        (v: ITableWithSortState) => v.tableId === id && _.isBoolean(v.isAsc),
    );

    const perPage =
        parseInt(
            FlatSelectSelectors.getSelectedOptionId(state, {
                id: PaginationUtils.getPaginationPerPageId(id),
            }),
            10,
        ) || undefined;

    const filter: IFilterState = _.findWhere(state.filters, {id});
    const predicates = getTablePredicates(id, state);

    return {
        // predicates
        predicates,

        // sort
        sortKey: tableSort && tableSort.id,
        sortAscending: (tableSort && tableSort.isAsc) || null,

        // pagination
        perPage,
        pageNb: _.findWhere(state.paginationComposite, {id: getPaginationId(id)})?.pageNb,

        // filter
        filter: filter && filter.filterText,

        // date
        dateLimits: DatePickerSelectors.getDatePickerLimits(state, {id}),
    };
};

/** e.g. tableId--componentId => tableId */
const getTableIdFromPredicateId = (predicateId: string) => predicateId.split(PREDICATE_SEPARATOR)[0];

/** e.g. tableId--componentId => componentId */
const getComponentIdFromPredicateId = (predicateId: string) => predicateId.split(PREDICATE_SEPARATOR)[1];

const getPredicateId = (tableId: string, componentId: string) => `${tableId}${PREDICATE_SEPARATOR}${componentId}`;

const getPredicateIds = (tableId: string, state: PlasmaState): string[] =>
    _.chain(state.listBoxes)
        .filter(filterTablePredicate.bind(null, tableId))
        .pluck('id')
        .map(getComponentIdFromPredicateId)
        .value();

const getPaginationId = (tableId: string) => `pagination-${tableId}`;

const getTablePredicates = (tableId: string, state: PlasmaState): ITableHOCPredicateValue[] =>
    _.chain(state.listBoxes)
        .filter(filterTablePredicate.bind(null, tableId))
        .filter((list: IListBoxState) => list.selected && list.selected[0] !== '')
        .map((list: IListBoxState) => ({id: getComponentIdFromPredicateId(list.id), value: list.selected[0]}))
        .value();

const getDatePickerId = (tableId: string) => `${tableId}-date-range`;

export const isRowCollapsible = (props: ITableRowOwnProps): boolean =>
    props.collapsible &&
    (isValidElement(props.collapsible.content) ||
        _.isString(props.collapsible.content) ||
        _.isNull(props.collapsible.content));

export const TableHOCUtils = {
    getCompositeState,
    getPredicateId,
    getPredicateIds,
    getPaginationId,
    getTableIdFromPredicateId,
    getDatePickerId,
};

const filterTablePredicate = (tableId: string, {id}: IListBoxState): boolean => new RegExp(`^${tableId}(.+)`).test(id);
