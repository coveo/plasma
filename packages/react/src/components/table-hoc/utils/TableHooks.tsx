import {CSSProperties, useRef, useState, useLayoutEffect} from 'react';

export const useFixedWidthWhileLoading = (isLoading: boolean) => {
    const tableHeaderRef = useRef<HTMLTableHeaderCellElement>();
    const [columnWidth, setColumnWidth] = useState<number>();

    useLayoutEffect(() => {
        if (tableHeaderRef.current && !isLoading) {
            setColumnWidth(tableHeaderRef.current.offsetWidth);
        }
    }, [isLoading]);

    const style: CSSProperties = isLoading
        ? {
              width: `${columnWidth}px`,
          }
        : {};

    return {style, tableHeaderRef};
};
