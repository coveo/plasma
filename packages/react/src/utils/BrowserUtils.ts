// https://stackoverflow.com/questions/21825157/internet-explorer-11-detection#answer-22242528
const isIE = () =>
    typeof window !== 'undefined' &&
    (navigator.userAgent.indexOf('MSIE') !== -1 || navigator.appVersion.indexOf('Trident/') > -1);

export const BrowserUtils = {
    isIE,
};
