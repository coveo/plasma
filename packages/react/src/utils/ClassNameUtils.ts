import * as _ from 'underscore';
export type ClassNameTypes = string | string[] | {[key: string]: boolean};

/**
 * Type to use when typing a property or variable that will be used with
 * the "classname" library (https://www.npmjs.com/package/classnames)
 */
export type IClassName = ClassNameTypes | ClassNameTypes[];

export interface IAdditionalClass {
    className: string;
    condition?: (data: any) => boolean;
}

export const getAdditionalClasses = (additionalClasses: IAdditionalClass[], dataToUse: any): IClassName =>
    _.map(additionalClasses, (additionalClass: IAdditionalClass) =>
        additionalClass.condition
            ? {
                  [additionalClass.className]: additionalClass.condition(dataToUse),
              }
            : additionalClass.className,
    );
