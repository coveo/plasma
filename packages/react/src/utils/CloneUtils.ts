export const deepClone = (object: any) => JSON.parse(JSON.stringify(object));
