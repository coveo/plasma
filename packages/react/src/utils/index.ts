export * from './BrowserUtils';
export * from './ClassNameUtils';
export * from './CloneUtils';
export * from './ComponentUtils';
export * from './DataStructuresUtils';
export * from './DateUtils';
export * from './FalsyValuesUtils';
export * from './FilterUtils';
export * from './HocUtils';
export * from './InputUtils';
export * from './JSONUtils';
export * from './JSXUtils';
export * from './PropsToOmitUtils';
export * from './RandomUtils';
export * from './ReduxUtils';
export * from './StringUtils';
export * from './TooltipUtils';
export * from './TypesUtils';
export * from './UrlUtils';
export * from './UUID';
export * from './ValidComponentChildren';
export * from './TypeUtils';
