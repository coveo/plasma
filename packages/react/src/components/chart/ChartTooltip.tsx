import classNames from 'classnames';
import {FunctionComponent, CSSProperties, ReactNode, useRef, useContext, useState, Fragment} from 'react';

import {DropPodPosition, IDropUIPosition} from '../drop/DomPositionCalculator';
import {DropPod} from '../drop/DropPod';
import {ChartTooltipContent} from './ChartTooltipContent';
import {XYChartContext, XYPoint} from './XYChart';

export interface ChartTooltipProps {
    sort?: boolean;
}

/**
 * @deprecated Use Mantine instead
 */
export const ChartTooltip: FunctionComponent<ChartTooltipProps> = ({sort = false}) => {
    const ref = useRef<HTMLElement>(null);
    const {series, xScale, yScale, xDomain, yDomain, width, height} = useContext(XYChartContext);
    const [position, setPosition] = useState({
        x: 0,
        y: 0,
        index: 0,
        position: null,
        pointX: null,
    });

    const onMouseMove = (index: number, pointX: number, x: number) => {
        const dropPodPosition = x < width / 2 ? DropPodPosition.right : DropPodPosition.left;
        setPosition({x, y: height / 2, index, pointX, position: dropPodPosition});
    };
    const onMouseLeave = () => setPosition({...position, position: null});

    const barWidth =
        series[0].data.length > 1
            ? (xScale(xDomain[1]) - xScale(xDomain[0])) / series[0].data.length
            : xScale(xDomain[0]);
    const bars = series[0].data.map((point: XYPoint, index: number) => {
        const x = xScale(point.x);
        return (
            <Fragment key={`chart-zone-${index}`}>
                <rect
                    fill="transparent"
                    width={barWidth / 2}
                    height={yScale(yDomain[0])}
                    x={x - barWidth / 2}
                    y={yScale(yDomain[1])}
                    onMouseMove={() => onMouseMove(index, point.x, x)}
                    onMouseLeave={onMouseLeave}
                />
                <rect
                    fill="transparent"
                    width={barWidth / 2}
                    height={yScale(yDomain[0])}
                    x={x}
                    y={yScale(yDomain[1])}
                    onMouseMove={() => onMouseMove(index, point.x, x)}
                    onMouseLeave={onMouseLeave}
                />
            </Fragment>
        );
    });
    return (
        <g className="chart-tooltip-zones">
            {!!position.position && (
                <rect
                    ref={ref as any}
                    className="chart-tooltip-line"
                    width={2}
                    x={xScale(position.pointX) - 1}
                    y={yScale(yDomain[1])}
                    height={yScale(yDomain[0])}
                    fill={'var(--deprecated-orange)'}
                />
            )}
            <circle cx={position.x} cy={position.y} r="1" fill="transparent" />
            <DropPod
                ref={ref}
                isOpen={!!position.position}
                positions={[position.position, DropPodPosition.left, DropPodPosition.right]}
                renderDrop={(style: CSSProperties, dropPosition: IDropUIPosition | null): ReactNode => (
                    <div
                        style={{...style, pointerEvents: 'none'}}
                        className={classNames('show-on-top', {
                            'ml-2': dropPosition?.position === DropPodPosition.left,
                            ml2: dropPosition?.position === DropPodPosition.right,
                        })}
                    >
                        <ChartTooltipContent x={position.index} sort={sort} />
                    </div>
                )}
            />
            {bars}
        </g>
    );
};
ChartTooltip.displayName = 'ChartTooltip';
