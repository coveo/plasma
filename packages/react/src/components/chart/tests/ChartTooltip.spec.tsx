import userEvent from '@testing-library/user-event';
import {render, screen} from '../../../../jest/utils';
import {ChartTooltip} from '../ChartTooltip';
import {XYChartContext} from '../XYChart';
import {XYChartContextMock} from './XYChartContextMock';

describe('<ChartTooltip />', () => {
    it('renders a tooltip when the user hover rectangles', async () => {
        const user = userEvent.setup();
        const {container} = render(
            <XYChartContext.Provider value={XYChartContextMock}>
                <svg role="application" width={XYChartContextMock.width} height={XYChartContextMock.height}>
                    <ChartTooltip />
                </svg>
            </XYChartContext.Provider>
        );

        expect(screen.getByRole('application')).toMatchSnapshot();

        // eslint-disable-next-line
        await user.hover(container.querySelectorAll('rect')[0]);

        expect(screen.getByText('First')).toBeVisible();

        // eslint-disable-next-line
        await user.unhover(container.querySelectorAll('rect')[0]);

        expect(screen.queryByText('First')).not.toBeVisible();
    });
});
