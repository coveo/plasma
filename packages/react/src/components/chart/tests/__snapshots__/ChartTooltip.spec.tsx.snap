// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChartTooltip /> renders a tooltip when the user hover rectangles 1`] = `
<svg
  height="50"
  role="application"
  width="10"
>
  <g
    class="chart-tooltip-zones"
  >
    <circle
      cx="0"
      cy="0"
      fill="transparent"
      r="1"
    />
    <rect
      fill="transparent"
      height="50"
      width="0.8333333333333334"
      x="-0.8333333333333334"
      y="0"
    />
    <rect
      fill="transparent"
      height="50"
      width="0.8333333333333334"
      x="0"
      y="0"
    />
    <rect
      fill="transparent"
      height="50"
      width="0.8333333333333334"
      x="1.1666666666666665"
      y="0"
    />
    <rect
      fill="transparent"
      height="50"
      width="0.8333333333333334"
      x="2"
      y="0"
    />
    <rect
      fill="transparent"
      height="50"
      width="0.8333333333333334"
      x="3.1666666666666665"
      y="0"
    />
    <rect
      fill="transparent"
      height="50"
      width="0.8333333333333334"
      x="4"
      y="0"
    />
    <rect
      fill="transparent"
      height="50"
      width="0.8333333333333334"
      x="5.166666666666667"
      y="0"
    />
    <rect
      fill="transparent"
      height="50"
      width="0.8333333333333334"
      x="6"
      y="0"
    />
    <rect
      fill="transparent"
      height="50"
      width="0.8333333333333334"
      x="7.166666666666667"
      y="0"
    />
    <rect
      fill="transparent"
      height="50"
      width="0.8333333333333334"
      x="8"
      y="0"
    />
    <rect
      fill="transparent"
      height="50"
      width="0.8333333333333334"
      x="9.166666666666666"
      y="0"
    />
    <rect
      fill="transparent"
      height="50"
      width="0.8333333333333334"
      x="10"
      y="0"
    />
  </g>
</svg>
`;
